[linker] Refactor and simplify code optimizations.

rolfbjarne opened the pull request:
We had several types of code optimizations, where we'd determine a particular
expression is a constant, then remove the condition and the code for any dead
branches.

Unfortunately each type of expression had its own logic to determine the
sequence of code to remove, each slightly different depending on the actual
conditional expression. This has led to bugs in the past, either when
switching between compilers (mcs/csc), or compilers have changed their emitted
code.

So implement a more generic version where each optimization just changes the
specific condition to a constant value, and then at the end we have a dead-
code-elimination pass that eliminates dead code based on those constant
conditions.

Additionally it will remove NOP instructions at the end of methods, because it's not verifiable IL, PEVerify complains:

    [IL]: Error: [D:\Documentos\Rolf\Xamarin\Xamarin.iOS.dll : AVFoundation.AVAssetImageGenerator::get_AppliesPreferredTrackTransform][offset 0x0000001E] fall through end of the method without returning

The result is that PEVerify shows much fewer errors, [1721][1] vs [2581][2].

This version is also a lot more defensive than the previous code, it will bail
out without doing anything if it finds a code sequence it doesn't understand.

This also makes it easier to implement other similar code optimizations.

Simple time measuring shows no slowdown in the linker, and the size
difference for monotouch-test is negligable (iPhone/Debug32):

32 bytes less for the executable:

    -rwxr-xr-x  1 rolf  staff  72890880 Jan 12 00:54:49 2018 /Users/rolf/test/old/monotouchtest.app/monotouchtest*
    -rwxr-xr-x  1 rolf  staff  72890848 Jan 12 00:56:57 2018 /Users/rolf/test/new/monotouchtest.app/monotouchtest*

12288 bytes less for Xamarin.iOS.dll (this is a debug build, a release build
would probably not show any difference at all):

    -rw-r--r--  1 rolf  staff  2506240 Jan 12 00:54:49 2018 /Users/rolf/test/old/monotouchtest.app/Xamarin.iOS.dll
    -rw-r--r--  1 rolf  staff  2493952 Jan 12 00:56:57 2018 /Users/rolf/test/new/monotouchtest.app/Xamarin.iOS.dll

[IL diff][3]

[1]: https://gist.github.com/rolfbjarne/9c8fa519a6f8e1718b125472f05ded07#file-peverify-new-txt-L1726
[2]: https://gist.github.com/rolfbjarne/9c8fa519a6f8e1718b125472f05ded07#file-peverify-old-txt-L2586
[3]: https://gist.github.com/rolfbjarne/35d5bfec1809ad2a80a7781cebe5b574

monojenkins commented:
Build failure

VincentDondain commented:
Code looks good to me (with my limited expertise of it).

I was just thinking that, while we have `OptimizeGeneratedCodeTest` and it's hard to test IL re-writing, it seems we don't cover some optimizations like `remove calls to UIApplication::EnsureUIThread` or `inline IsDirectBinding`.

spouliot commented:
build

spouliot commented:
@VincentDondain that's covered by some tests (e.g. LinkUiThread.cs) but running our tests with the linker enabled is even better. The main issue (not fixable by adding tests) is that a custom compiler might not generate the same code (which is why we only optimize code we know about).

monojenkins commented:
Build success

spouliot commented:
https://gist.github.com/rolfbjarne/33590ad651a21f9a9352ac9b97b9dc06#file-foo-diff-L336
That's not correct, a method should end with `ret` not `nop`

spouliot commented:
That's unexpected -> MonoTouchAssembly -> PlatformAssembly
https://gist.github.com/rolfbjarne/33590ad651a21f9a9352ac9b97b9dc06#file-foo-diff-L1621

spouliot commented:
I wonder if changes like

```
     call       uint64 ObjCRuntime.Messaging::UInt64_objc_msgSendSuper(native int,
                                                                       native int)
     stloc.0
-    br         IL_005a
-
+    nop
     nop
     nop
     nop
```

could force the JIT/AOT compiler to use long branches (instead of short ones) ? maybe that would explain the small size increase...

spouliot commented on tools/linker/CoreOptimizeGeneratedCode.cs:
```diff
@@ -80,5 +81,362 @@ static protected void Nop (Instruction ins)
+			Console.WriteLine ();
+#endif
+
+			// Exterminate, exterminate, exterminate
```
well done dr :)

spouliot commented on tools/linker/MonoTouch.Tuner/OptimizeGeneratedCodeSubStep.cs:
```diff
@@ -68,198 +69,154 @@ protected override void Process (TypeDefinition type)
-			if (fr.Name != field)
-				return false;
-			return fr.DeclaringType.Is (nspace, type);
+			return;
```
needed ?

spouliot replied:
no (but minor)

rolfbjarne commented:
@spouliot 

> That's unexpected -> MonoTouchAssembly -> PlatformAssembly https://gist.github.com/rolfbjarne/33590ad651a21f9a9352ac9b97b9dc06#file-foo-diff-L1621

There are two statements that changed order: 
https://github.com/xamarin/xamarin-macios/blob/07c24949c0b846cd1e65f77d13c68dc34f3ad496/src/Foundation/NSObject2.cs#L66 and https://github.com/xamarin/xamarin-macios/blob/07c24949c0b846cd1e65f77d13c68dc34f3ad496/src/Foundation/NSObject.iOS.cs#L23

Since these are static fields in two different files (partial class), I believe the compiler is free to execute the statements in either order, so this is just (unfortunate) noise.

rolfbjarne commented:
@spouliot I've addressed your comments, and I've also updated the PR description with updated info after running tests again (app is now smaller, and PEVerify is happier).

monojenkins commented:
Build failure

rolfbjarne commented:
Test failure is unrelated:

* mmptest: https://github.com/xamarin/maccore/issues/612
